lib/open_api_spex/cast/discriminator.ex

```diff
@@ -0,0 +1,104 @@
+defmodule OpenApiSpex.Cast.Discriminator do
+  @moduledoc """
+  Defines the `OpenApiSpex.Discriminator.t` type.
+  """
+
+  alias OpenApiSpex.{Cast, Reference}
+
+  @enforce_keys :propertyName
+  defstruct [
+    :propertyName,
+    :mapping
+  ]
+
+  @typedoc """
+  [Discriminator Object](https://swagger.io/specification/#discriminatorObject)
+
+  When request bodies or response payloads may be one of a number of different schemas,
+  a discriminator object can be used to aid in serialization, deserialization, and validation.
+  The discriminator is a specific object in a schema which is used to inform the consumer of the
+  specification of an alternative schema based on the value associated with it.
+
+  A discriminator requires a composite key be set on the schema:
+
+    * `allOf`
+    * `oneOf`
+    * `anyOf`
+  """
+  @type t :: %__MODULE__{
+          propertyName: String.t(),
+          mapping: %{String.t() => %Reference{} | String.t()} | nil
+        }
+
+  def cast(ctx) do
+    case cast_discriminator(ctx) do
+      {:ok, result} -> {:ok, result}
+      error -> error
+    end
+  end
+
+  defp cast_discriminator(%_{value: value, schema: schema} = ctx) do
+    {discriminator_property, mappings} = discriminator_details(schema)
+
+    case Map.pop(value, "#{discriminator_property}") do
+      {"", _} ->
+        error(:no_value_for_discriminator, ctx)
+
+      {discriminator_value, castable_value} ->
+        # The cast specified by the composite key (allOf, anyOf, oneOf) MUST succeed
+        # or return an error according to the Open API Spec.
+        composite_ctx = %{
+          ctx
+          | value: castable_value,
+            schema: %{schema | discriminator: nil},
+            path: ["#{discriminator_property}" | ctx.path]
+        }
+
+        with {composite_schemas, {:ok, _}} <- cast_composition(composite_ctx),
```

Comment:
This function is getting pretty deep, probably due for a refactor.

Reply:
The refactor on this function is the only fix remaining I think... I'll ping you @bryannaegele when I commit it.